File: dongyoung/chap8/boj/1932.cpp
/**
* BOJ 1932 정수 삼각형 실버1
 * 입력: n(1 ~ 500), 정수 삼각형
 * 출력: 합이 최대가 되는 경로에 있는 수의 합
*/
#include <bits/stdc++.h>
using namespace std;

int n;
int input[501][501];
int d[501][501];

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    cin >> n;
    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= i; j++) {
            cin >> input[i][j];
        }
    }
    d[1][1] = input[1][1];
    d[2][1] = input[1][1] + input[2][1];
    d[2][2] = input[1][1] + input[2][2];
    for (int i = 3; i <= n; i++) {
        for (int j = 1; j <= i; j++) {
            d[i][j] = max(d[i-1][j-1], d[i-1][j]) + input[i][j];
        }
    }
    int mx = d[n][1];
    for (int i = 1; i <= n; i++) {
        mx = max(mx, d[n][i]);
    }
    cout << mx;

}

File: woogieon8on/chap8/boj/1932.py
n = int(input())

tree = []
for _ in range(n):
    tree.append(list(map(int, input().split())))

for i in range(1, n): # 첫 번째 행 제외
    for j in range(len(tree[i])):
        if j == 0: # 첫 번째 열의 경우 이전 행의 첫 번째 열 선택
            tree[i][j] = tree[i][j] + tree[i-1][j]
        elif j == len(tree[i]) - 1: # 마지막 열의 경우 이전 행의 마지막 열 선택
            tree[i][j] = tree[i][j] + tree[i-1][j-1]
        else: # 나머지는 이전 행의 같은 열, (자신의 열 - 1)에 해당하는 값 중에서 최대값 선택
            tree[i][j] = tree[i][j] + max(tree[i-1][j-1], tree[i-1][j])

print(max(tree[n-1])) # 마지막 행에서 최대값 출력